Some micro optimizations for `qdel()`

SyncIt21 opened the pull request:
## About The Pull Request
How did this not runtime 1 decade ago or something? And if it hasn't then do we even need this null check? (maybe remove it save some micro ticks or something?), cause if we do then we need to move it up to the top of the proc so we can early return and save processing. Or something like that idk this makes sense right?

The `istype()` check below will pass if `to_delete` is null and pass `null` to `del()` which also doesn't seem ideal so this should work better?

I also moved up some other checks before we start creating instances of `datum/qdel_item` so you know early return & save time if the object is already getting deleted? Again i don't know someone please advice.

Also stored the type for `to_delete` inside a var so you save some memory access time for datum variables

## Changelog
N/A

SyncIt21 committed:
move up?

SyncIt21 committed:
these are pure operations so should be okay?

SyncIt21 committed:
Merge branch 'master' into qdelNullUhh

SyncIt21 committed:
maybe this speeds things up

Kapu1178 reviewed (commented):
Nullchecking before just adds more work to the hottest path. I think overall this is a speed loss. del(null) silently fails and thats okay.

Setting type to a var is also slower for the hottest path because its only used in extremely rare paths.

Mothblocks commented:
+1 to Kapu--I am skeptical the null check is valuable here.

Furthermore, caching the type is also not particularly faster--every CRASH is a cold path, the hottest path here is qdeling something that's already been qdel'd, in which case we only use the type once.

I don't want this without benchmarks to show it's faster, because my hunch is it is not.

SyncIt21 commented:
I can remove the null check that's fine But you forgot to review the other changes.

> I also moved up some other checks before we start creating instances of datum/qdel_item so you know early return & save time if the object is already getting deleted?

I moved up the crash check and signal checks before we create the datum/qdel_item, what about that?
